### ISSUE TITLE
kernel: [297410.332718] BTRFS error (device bcache3): bdev /dev/bcache3 errs: wr 0, rd 1, flush 0, corrupt 0, gen 0

### ISSUE THREAD

**Issue by kakra** [Zygo/bees#56]
When running bees, I see the following kernel message from time to time:

    [297410.332718] BTRFS error (device bcache3): bdev /dev/bcache3 errs: wr 0, rd 1, flush 0, corrupt 0, gen 0

This is very very rare, with increasing rd errors in timescales of months.

However, all the files seem to be okay, HDD surfaces are okay (according to smartctl). There are no other errors in the kernel log around these messages.

Using journalctl I am able to get this in context with the bees logs:

```
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0x20000..0x40000] {0x8970578000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0x20000..0x40000] {0x897090c000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0x20000 [Dddddddddddddddddddddddddddddddd] 0x40000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 20K [0x9db55f000..0x9db564000] {0x2adc78f000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 20K [0x979ee4000..0x979ee9000] {0x291627e000z19000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 20K 0x979ee4000 [DDDdd] 0x979ee9000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0x40000..0x60000] {0x8970580000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0x40000..0x60000] {0x8970914000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0x40000 [Dddddddddddddddddddddddddddddddd] 0x60000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: WORKAROUND: abandoned toxic match for hash 0x88ae7d417641951e addr 0x29151e5000tz1e000
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0x60000..0x80000] {0x8970585000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0x60000..0x80000] {0x8970919000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0x60000 [Dddddddddddddddddddddddddddddddd] 0x80000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: WORKAROUND: abandoned toxic match for hash 0x7bdb4b8d1a2774d1 addr 0x2a7fe95000t
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0x80000..0xa0000] {0x897058a000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0x80000..0xa0000] {0x897091e000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0x80000 [Dddddddddddddddddddddddddddddddd] 0xa0000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: dedup: src 12K [0x9b7099000..0x9b709c000] {0x2915780000z0} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl:        dst 12K [0x979eeb000..0x979eee000] {0x2960aaf000z0} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:56 jupiter beesd[12654]: tid 12711 crawl: scan: 12K 0x979eeb000 [DDD] 0x979eee000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter kernel: BTRFS error (device bcache3): bdev /dev/bcache3 errs: wr 0, rd 1, flush 0, corrupt 0, gen 0
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0xa0000..0xc0000] {0x8970590000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0xa0000..0xc0000] {0x8970924000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0xa0000 [Dddddddddddddddddddddddddddddddd] 0xc0000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 32K [0x9ab9a9000..0x9ab9b1000] {0x290fed4000z3000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 32K [0x979eee000..0x979ef6000] {0x2917d02000z3000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 32K 0x979eee000 [DDdddddd] 0x979ef6000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 128K [0xc0000..0xe0000] {0x897059b000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 128K [0xc0000..0xe0000] {0x897092f000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0xc0000 [Dddddddddddddddddddddddddddddddd] 0xe0000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 32K [0x9b7a64000..0x9b7a6c000] {0x2915d64000zb000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 32K [0x979ef6000..0x979efe000] {0x2917b80000zb000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 32K 0x979ef6000 [DDdddddd] 0x979efe000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 100K [0xe0000..0xf9000] {0x5575ef000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 100K [0xe0000..0xf9000] {0x897093a000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 12K [0x110000..0x113000] {0x89707d0000z13000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 12K [0xfd000..0x100000] {0x897093a000z1d000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: copy: 16K [0xf9000..0xfd000] fd = 112 'home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim'
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 16K [0xc2d6000..0xc2da000] {0xd02550000z0} #3208 (deleted)
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 16K [0xf9000..0xfd000] {0x897093a000z19000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0xe0000 [ddddddddddddddddddddddddd++++ddd] 0x100000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 16K [0x9b9a7f000..0x9b9a83000] {0x29177c6000z13000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 16K [0x979efe000..0x979f02000] {0x2917f06000z13000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 16K 0x979efe000 [DDdd] 0x979f02000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 60K [0x113000..0x122000] {0x89707d0000z16000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt_to_idle_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 60K [0x100000..0x10f000] {0x897093e000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 16K [0x110000..0x114000] {0x89705aa000z1000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 16K [0x10f000..0x113000] {0x897093e000zf000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 52K [0xff000..0x10c000] {0x89705aa000z5000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 52K [0x113000..0x120000] {0x897093e000z13000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 128K 0x100000 [DddddddddddddddDdddDdddddddddddd] 0x120000 home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 36K [0x9b6290000..0x9b6299000] {0x2915108000z17000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 36K [0x979f02000..0x979f0b000] {0x2917fe8000z17000} home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: scan: 36K 0x979f02000 [DDddddddd] 0x979f0b000 home/kakra/.VirtualBox/VDI/WindowsXPPro_C.vdi
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl: dedup: src 32K [0x10c000..0x114000] {0x89705aa000z12000} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_hurt.dae_anim
Jan 18 03:47:57 jupiter beesd[12654]: tid 12711 crawl:        dst 32K [0x120000..0x128000] {0x897093f000z0} home/kakra/.local/share/Steam/SteamApps/common/SOMA/entities/character/robot/kate_soma/animations/move_to_idle.dae_anim
```

Maybe bees triggers another kernel bug?

**Comment by Zygo** [comment]
Not a bug I've seen.

What is causing the rd count to increase?  There should be other kernel log messages when the error is detected (both from btrfs and lower layer storage).

**Comment by kakra** [comment]
No, there's never another log. Just this message.

**Comment by Zygo** [comment]
I did a test run with bees + bcache on 4.14.x a month or two ago.  It didn't end well--22 hours in, the kernel locked up and the filesystem was damaged (nothing a scrub couldn't fix).  lvmcache had similar problems but lasted for several weeks before failing.  The SSD hardware I used for the test ran bees without a caching block device for a year before, and a month since, without incident.

A user reported similar problems with bcache that stopped after upgrading to 4.15.

I'm hesitant to close this without more information, but if no more information arrives, then there's nothing to be done about this issue.

**Comment by kakra** [comment]
Currently, I no longer see the issue. But I'm also using ewheeler's patches which bypass bcache for low io-prio processes. And I've made bees into low io-prio. So the bug may still be there and may be connected to bcache - but I actually didn't see it since months.

**Comment by kakra** [comment]
Remembering the last list threads, it also looks like bcache fixed a lot of locking issues since 4.14. Maybe those should be back-ported? Or the patches I'm using have a real impact because bees won't be passed through bcache on writes or reads (except blocks are already in bcache).

**Comment by kakra** [comment]
Let's close this, not seen this again...
